src/StackExchange.Redis/RedisMetrics.cs

```diff
@@ -0,0 +1,79 @@
+﻿#if NET6_0_OR_GREATER
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.Metrics;
+
+namespace StackExchange.Redis;
+
+internal sealed class RedisMetrics
+{
+    private static readonly double s_tickFrequency = (double)TimeSpan.TicksPerSecond / Stopwatch.Frequency;
+    // cache these boxed boolean values so we don't allocate on each usage.
+    private static readonly object s_trueBox = true;
+    private static readonly object s_falseBox = false;
+
+    private readonly Meter _meter;
+    private readonly Counter<long> _operationCount;
+    private readonly Histogram<double> _messageDuration;
+    private readonly Counter<long> _nonPreferredEndpointCount;
+
+    public static readonly RedisMetrics Default = new RedisMetrics();
+
+    public RedisMetrics(Meter? meter = null)
+    {
+        _meter = meter ?? new Meter("StackExchange.Redis");
+
+        _operationCount = _meter.CreateCounter<long>(
+            "db.redis.operation.count",
+            description: "The number of operations performed.");
+
+        _messageDuration = _meter.CreateHistogram<double>(
+            "db.redis.duration",
+            unit: "s",
+            description: "Measures the duration of outbound message requests.");
+
+        _nonPreferredEndpointCount = _meter.CreateCounter<long>(
+            "db.redis.non_preferred_endpoint.count",
+            description: "Indicates the total number of messages dispatched to a non-preferred endpoint, for example sent to a primary when the caller stated a preference of replica.");
+    }
+
+    public void IncrementOperationCount(string endpoint)
+    {
+        _operationCount.Add(1,
+            new KeyValuePair<string, object?>("endpoint", endpoint));
+    }
+
+    public void OnMessageComplete(Message message, IResultBox? result)
+    {
+        // The caller ensures we can don't record on the same resultBox from two threads.
+        // 'result' can be null if this method is called for the same message more than once.
+        if (result is not null && _messageDuration.Enabled)
+        {
+            // Stopwatch.GetElapsedTime is only available in net7.0+
+            // https://github.com/dotnet/runtime/blob/ae068fec6ede58d2a5b343c5ac41c9ca8715fa47/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Stopwatch.cs#L129-L137
+            var now = Stopwatch.GetTimestamp();
+            var duration = new TimeSpan((long)((now - message.CreatedTimestamp) * s_tickFrequency));
+
+            var tags = new TagList
```

Comment:
for upto 3 tags, its faster to pass them directly.
https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/metrics#instruments
_When reporting measurements with 3 tags or less, pass the tags directly to the instrument API._